### PR TITLE
Bug 1343328 - [DO NOT MERGE] PerfDatum bigint PK migration using pt-osc

### PR DESCRIPTION
**Pull request by vrigal**
There are 3 migrations:
* perf.0043 drops the `perf_muulticommitdatum` table, allowing to do the `performance_datum` PK migration with an external tool. There is a check that requires the table to be empty for the migration to pass.
* perf.0044 update django's state schema. It requires `performance_datum` PK column to be of type `bigint(20)` to pass. This migration is not reversible and restore the `perf_multicommitdatum` table (handled by django, no error should occur there).
* perf.0045 restores a specific DB constraint. It is a copy of perf.0036 that updates  `perf_muulticommitdatum` FK constraint (to add a CASCADE behavior on delete).

There are checks to ensure migrations are not ran if the data is not ready on the tables at each step:
```sh
$ django-admin migrate perf 0043
$ #Perform the PK migration directly on the table
$ django-admin migrate perf
```

### CODE REVIEW

**Comment by kkleemola** [comment]
Thanks for this!

I think this should work although I haven't tested it yet. My only initial concern would be if we have any dev envs, etc where we might not have empty tables and then this migration fails without a clear way to resolve. I'm not sure if that's a scenario that exists with treeherder's development flow (y'all would know more than me).

In my head, I thought that we were using the django migration to drop perf_multicommitdatum in order to temporary disable the feature and avoid errors.

Since we're okay with some errors, we can probably simplify this further and use pt-osc's hooks feature to update perf_multicommitdatum. That would then simplify this migration to just the naive table alteration django would generate by updating the model. I think we could then just run django-admin migrate -fake name
after we run pt-osc and before this is merged and deployed so it wouldn't actually be performed on any of our environments.
Basically this would still ensure we have a simple migration for dev purposes and the mapping between model and db is updated . I think it might give a cleaner separation between the app and external tool. (i.e. the app is "unaware" we did a complicated schema migration under the hood.)

Otherwise I think this as is should be okay and I'll do a test run.

**Comment by kkleemola** [comment]
I did some tests as well as a successful run on prototype so this wfm. :+1:

**Review by kkleemola — APPROVED** [review]

**Comment by vrigal** [comment]
According to @kkleemola review, I'll split the second migration perf.0044 into two:
* `0044` will run the SQL to ALTER `performance_datum.id` to bigint. This one will be faked on staging/prod as the migration will be done with pt-osc
* `0045` will restore `perf_multicommitdatum` and update django schema

**Comment by kkleemola** [comment]
In my initial review what I meant by simplifying was instead have the migration simply be:

```
class Migration(migrations.Migration):

    dependencies = [
        ('perf', '0042_backfillrecord_new_fields'),
    ]

    operations = [
        migrations.AlterField(
            model_name='performancedatum',
            name='id',
            field=models.BigAutoField(primary_key=True, serialize=False),
        ),
    ]
```
which is what django generates with the model change.

The full migration would look something like this:

django-admin migrate -fake perf 0043
pt-osc modify column modify column id bigint not null auto_increment (on performance_datum)
(with plugin hooks for before and after the swap to deal with `perf_multicommitdatum`)

The PR could then be merged anytime after the migration has been faked with the caveat in release notes that there should be a warning that this migration shouldn't be performed on a large table (similar to what you have now). 

But like I said this works as well:
Your PR provides a bunch of safeguards so we don't accidentally deploy which is great.
But it might be fine to just make sure the caveats are documented in release notes and have a simple PR so we're not switching back and forth between tools.

I think this is largely a question to the maintainers on given the migration can't be easily be done as a django migration, how much of the migration logic is still necessary to have in a django migration?

**Comment by kkleemola** [comment]
@Archaeopteryx @jmaher Did either of you want to review this before I perform any migration work on stage?

**Review by jmaher — APPROVED** [review]

File: treeherder/perf/migrations/0044_perfdatum_bigint_fk.py
Comment (Archaeopteryx) [inline treeherder/perf/migrations/0044_perfdatum_bigint_fk.py:1]: `updates`

File: treeherder/perf/migrations/0045_restore_perf_multicommitdatum_and_schema.py
Comment (Archaeopteryx) [inline treeherder/perf/migrations/0045_restore_perf_multicommitdatum_and_schema.py:25]: `restores`

**Review by Archaeopteryx — APPROVED** [review]

**Comment by kkleemola** [comment]
The issue is once it's merged it will be difficult to do any production deploys until we do the production migration as the deployment pipeline will fail at the migrate step.
We were aiming for Tuesday to try and get this done. See: https://mozilla-hub.atlassian.net/browse/SVCSE-602

**Comment by jmaher** [comment]
we can wait a week or two if needed to deploy changes to treeherder production- I am deploying all the recent changes now to production

**Comment by vrigal** [comment]
I did not know there was an automatic deployment pipeline. It will causes an issue here because of the confirmation dialog (it must be ran manually or faked).
Also, `perf.0043` is required to perform the migration (it drops `perf_multicommitdatum` FK), but making a separate MR would break things (CI, and code related to MultiCommitDatum).

Is there any way to apply this patch on staging and manually handle the migration ? I don't know much about deployment.

**Comment by kkleemola** [comment]
Yes - this is going to be manually applied.
This is why I asked for [DO NOT MERGE]
Particularly given the last PR was accidentally merged which creates additional work for me (i.e. I need to manually modify the db)

**Comment by kkleemola** [comment]
Migration has been applied to stage.
Prod is scheduled for the 28th (Tuesday).

**Pull request merged**
